src/metal/rt64_metal.cpp

```diff
@@ -0,0 +1,2363 @@
+        state.depthClipMode = (desc.depthClipEnabled) ? MTL::DepthClipModeClip : MTL::DepthClipModeClamp;
+        state.winding = MTL::WindingClockwise;
+        state.renderPipelineState = device->mtl->newRenderPipelineState(descriptor, &error);
+
+        if (error != nullptr) {
+            fprintf(stderr, "MTLDevice newRenderPipelineState: failed with error %s.\n", error->localizedDescription()->utf8String());
+            return;
+        }
+
+        // Release resources
+        vertexDescriptor->release();
+        vertexFunction->release();
+        descriptor->release();
+        depthStencilDescriptor->release();
+        releasePool->release();
+    }
+
+    MetalGraphicsPipeline::~MetalGraphicsPipeline() {
+        if (state.renderPipelineState) state.renderPipelineState->release();
+        if (state.depthStencilState) state.depthStencilState->release();
+    }
+
+    RenderPipelineProgram MetalGraphicsPipeline::getProgram(const std::string &name) const {
+        assert(false && "Graphics pipelines can't retrieve shader programs.");
+        return RenderPipelineProgram();
+    }
+
+    // MetalDescriptorSet
+
+    MetalDescriptorSet::MetalDescriptorSet(MetalDevice *device, const RenderDescriptorSetDesc &desc) {
+        assert(device != nullptr);
+
+        this->device = device;
+
+        thread_local std::unordered_map<RenderDescriptorRangeType, uint32_t> typeCounts;
+        typeCounts.clear();
+
+        // Figure out the total amount of entries that will be required.
+        uint32_t rangeCount = desc.descriptorRangesCount;
+        if (desc.lastRangeIsBoundless) {
+            assert((desc.descriptorRangesCount > 0) && "There must be at least one descriptor set to define the last range as boundless.");
+
+            // Ensure at least one entry is created for boundless ranges.
+            const uint32_t boundlessRangeSize = std::max(desc.boundlessRangeSize, 1U);
+
+            const RenderDescriptorRange &lastDescriptorRange = desc.descriptorRanges[desc.descriptorRangesCount - 1];
+            typeCounts[lastDescriptorRange.type] += boundlessRangeSize;
+            rangeCount--;
+        }
+
+        // Spirv-cross orders by binding number, so we sort
+        std::vector sortedRanges(desc.descriptorRanges, desc.descriptorRanges + desc.descriptorRangesCount);
+        std::sort(sortedRanges.begin(), sortedRanges.end(), [](const RenderDescriptorRange &a, const RenderDescriptorRange &b) {
+            return a.binding < b.binding;
+        });
+
+        for (uint32_t i = 0; i < rangeCount; i++) {
+            const RenderDescriptorRange &descriptorRange = sortedRanges[i];
+            typeCounts[descriptorRange.type] += descriptorRange.count;
+        }
+
+        setLayout = std::make_unique<MetalDescriptorSetLayout>(device, desc);
+
+        argumentBuffer = {
+                .mtl = device->mtl->newBuffer(DESCRIPTOR_RING_BUFFER_SIZE, MTL::ResourceStorageModeManaged),
```

Comment:
What's the reasoning behind the "ring buffer" logic? I didn't see anything that ends up with the buffer being used as one. Should it not be based on the amount of descriptors instead of allocating a 1 MB fixed chunk?